From b3da66def63f671e75b3dbe51654ae2975c1987a Mon Sep 17 00:00:00 2001
From: PRESFIL <PRESFIL@users.noreply.github.com>
Date: Thu, 12 Jan 2023 11:39:09 +0000
Subject: [PATCH 1/2] getStackSubdir: support $STACK_ROOT env variable

Add support of `$STACK_ROOT` environment variable to follow `stack`'s
behaviour. It allows to override the Stack Root location.

Ref: https://docs.haskellstack.org/en/stable/environment_variables/#stack_root
---
 ChangeLog.md       |  4 ++++
 README.md          | 10 +++++++---
 src/Directories.hs | 13 +++++++++++--
 3 files changed, 22 insertions(+), 5 deletions(-)

diff --git a/ChangeLog.md b/ChangeLog.md
index 4b850d3..028ee83 100644
--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -1,5 +1,9 @@
 # Release history for stack-clean-old
 
+## Unreleased
+- add support for `${STACK_ROOT}` environment variable to override the Stack
+  Root location
+
 ## 0.4.7 (2023-08-22)
 - add --tarballs for programs/ ghc tarballs too (#9)
 - bug fix to match both ghc-*.* and ghc-variant-*.* (#14)
diff --git a/README.md b/README.md
index 29f6863..98d5c87 100644
--- a/README.md
+++ b/README.md
@@ -6,13 +6,17 @@ snapshot builds and ghc versions to recover diskspace.
 ## Usage
 `stack-clean-old [size|list|remove|keep-minor|purge-older|delete-work] [(-P|--project)|(-G|--global)|(-C|--compilers)|(-T|--tarballs)] [(-s|--subdirs)|(-r|--recursive)] [-d|--delete] [GHCVER]`
 
-In a project directory it acts on `.stack-work/install/` by default,
-otherwise on `~/.stack/{snapshots,programs}/`.
+In a project directory it acts on `.stack-work/install/` by default, otherwise
+on the *Stack Root location* `${STACK_ROOT}/{snapshots,programs}/`. If
+`${STACK_ROOT}` is unset, the *Stack Root location* is `~/.stack/`. See
+[official documentation][stack_root].
+
+[stack_root]: https://docs.haskellstack.org/en/stable/environment_variables/#stack_root
 
 Subcommands:
 
 `size`:
-    prints the total size of `.stack-work/` or `~/.stack/`
+    prints the total size of `.stack-work/` of project(s) or the *Stack Root location*.
     (`size` does not take a GHCVER argument).
 
 `list`:
diff --git a/src/Directories.hs b/src/Directories.hs
index 5b2cce2..eb015a3 100644
--- a/src/Directories.hs
+++ b/src/Directories.hs
@@ -9,6 +9,7 @@ where
 import Data.List.Extra
 import SimpleCmd (error')
 import System.Directory
+import System.Environment
 import System.FilePath
 import System.FilePath.Glob
 
@@ -16,10 +17,18 @@ globDirs :: String -> IO [FilePath]
 globDirs pat = do
   map (dropPrefix "./") <$> namesMatching (pat ++ "/")
 
+getStackRootDir :: IO FilePath
+getStackRootDir = do
+  home <- getHomeDirectory
+  env <- lookupEnv "STACK_ROOT"
+  case env of
+    Just path | isAbsolute path -> return path
+    _                           -> return $ home </> ".stack"
+
 getStackSubdir :: FilePath -> IO FilePath
 getStackSubdir subdir = do
-  home <- getHomeDirectory
-  return $ home </> ".stack" </> subdir
+  stackRoot <- getStackRootDir
+  return $ stackRoot </> subdir
 
 switchToSystemDirUnder :: Maybe String -> FilePath -> IO ()
 switchToSystemDirUnder msystem dir = do
-- 
2.42.0

